<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Jiadong Mao" />

<meta name="date" content="2025-05-22" />

<title>Annotation of scATAC-seq using scRNA-seq reference</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Annotation of scATAC-seq using scRNA-seq
reference</h1>
<h4 class="author">Jiadong Mao</h4>
<h4 class="date">2025-05-22</h4>



<div id="highlights" class="section level2">
<h2>Highlights</h2>
<ul>
<li><p>ScRNA-seq reference + scATAC-seq query</p></li>
<li><p>Utilising a bimodal bridge dataset</p></li>
<li><p>Comprehensive benchmark study</p></li>
</ul>
</div>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>Single-cell ATAC-seq (Assay for Transposase-Accessible Chromatin
sequencing) measures chromatin accessibility across the genome. In the
nucleus of each cell, each DNA chain is like a long thread wrapping
around protein spools called histones. Each DNA-wrapped histone is known
as a nucleosome – the basic unit of chromatin. Measuring the openness of
chromatin regions tells us two things:</p>
<ul>
<li><p>Whether a gene can be transcribed. For a gene to be transcribed,
the gene’s corresponding chromatin region has to be open, i.e. part of
the DNA chain being unwound.</p></li>
<li><p>Transcription factor (TF) footprints. TFs bind to gene regions to
promote or suppress gene expression. ATAC-seq can infer which regions
TFs are binding to.</p></li>
</ul>
<p>ScATAC-seq allows us to infer cell-type-specific gene regulations.
Hence identifying cell types based on scATAC-seq is an important step.
One common appraoch is to view ATAC-seq as surrogate for RNA-seq. That
is, if a gene has more open regions, then we assume that it has more
transcripts.</p>
<p>In this case study, we use a reference-based approach to annotate
cell types in scATAC-seq. This approach is inspired by bridge
integration implemented in Seurat V5 (<a href="https://www.nature.com/articles/s41587-023-01767-y">Hao et al.,
2024</a>). The idea is to transfer cell type annotations in scRNA-seq
reference to scATAC-seq query, via a bimodal bridge dataset with matchcd
RNA and ATAC features (e.g. generated from 10x multiome).</p>
<p>We use</p>
<ul>
<li><p>Reference: BMMC (bone marrow mononuclear cell) scRNA-seq data set
(<a href="https://www.cell.com/cell/fulltext/S0092-8674(19)30559-8?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS0092867419305598%3Fshowall%3Dtrue">Stuart
et al. (2019)</a>);</p></li>
<li><p>Bridge and query: 10x multiome (scRNA+ATAC-seq) dataset with 13
batches from <a href="https://datasets-benchmarks-proceedings.neurips.cc/paper/2021/hash/158f3069a435b314a80bdcb024f8e422-Abstract-round2.html">Leucken
et al. (2021)</a>.</p></li>
</ul>
<p>In our <a href="https://www.biorxiv.org/content/10.1101/2024.06.19.599787v1.full">paper</a>
we used a cross-validation (CV) scheme as follows. Each time we use one
of the 13 batches of multiome data as the bridge dataset to facilitate
the transfer of cell type annotations from the scRNA-seq reference to
the remaining 12 batches of query scATAC-seq data (with their RNA part
and ground truth cell type labels hidden). This allowed us to benchmark
PhiSpace cross-modality annotation with Seurat V5. Here we simply show
how PhiSpace bridge annotation can be done using one bridge and one
query datasets.</p>
<p>All data related to this case study can be downloaded <a href="https://www.dropbox.com/scl/fo/jeuqzjfyyr2j7doa922ve/AI-2U_wtZBpPGOswxMzReGQ?rlkey=n328yyr2llf81gz3chynjg0r6&amp;st=nb1svufg&amp;dl=0">here</a>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(PhiSpace))</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(ggplot2))</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(dplyr))</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(magrittr))</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(ggpubr))</span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(tidyr))</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(RColorBrewer))</span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a>dat_dir <span class="ot">&lt;-</span> <span class="st">&quot;~/Dropbox/Research_projects/PhiSpace/VignetteData/ATAC/&quot;</span> <span class="co"># Replace this by your own directory </span></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a><span class="fu">source</span>(<span class="st">&quot;~/Dropbox/Research_projects/PhiSpace/VignetteData/ATAC/utils.R&quot;</span>)</span></code></pre></div>
</div>
<div id="peaks-or-aggregated-peaks" class="section level1">
<h1>Peaks or aggregated peaks?</h1>
<p>ScATAC-seq data are often available in the peak by cell format, where
a peak refers to a small region on the genome that are open in a
significant number of cells (i.e. region of interest). Peaking calling,
i.e. defining the peaks based on the raw sequencing data, is a
sophisticated computational step and there are a few different
approaches, which we cannot cover here. Peaks are often further
aggregated to the gene level as a proxy of gene expression. That is,
open peaks falling in the region of a certain gene are summed up as
‘gene activity score’, approximating the expression level of that
gene.</p>
<p>In our paper, we demonstrated that gene activite scores (aggregted
peaks) are more suitable for cell typing due to its enhanced data
quality. Here we show PhiSpace cross-modality cell typing using both
peaks and gene activity scores.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>reference <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(dat_dir, <span class="st">&quot;data/obj.rna_for_refMap_sce.rds&quot;</span>))</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>bridgeRNA <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(dat_dir, <span class="st">&quot;data/bridgeRNA_s4d8.rds&quot;</span>))</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>bridgeATACpeaks <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(dat_dir, <span class="st">&quot;data/bridgeATACpeaks_s4d8.rds&quot;</span>))</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a>bridgeGA <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(dat_dir, <span class="st">&quot;data/bridgeGA_s4d8.rds&quot;</span>))</span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>queryPeaks <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(dat_dir, <span class="st">&quot;data/queryATACpeaks_s1d1.rds&quot;</span>))</span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>queryGA <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(dat_dir, <span class="st">&quot;data/queryATAC_GA_s1d1.rds&quot;</span>))</span></code></pre></div>
<p>Run PhiSpace to transfer annotations from reference scRNA-seq to
bridge scRNA-seq (RNA part of the bimodal bridge dataset).</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>PhiResPath <span class="ot">&lt;-</span> <span class="fu">paste0</span>(dat_dir, <span class="st">&quot;output/PhiResRNA.rds&quot;</span>)</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">file.exists</span>(PhiResPath)){</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>  PhiSpaceAssay <span class="ot">&lt;-</span> <span class="st">&quot;logcounts&quot;</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>  YtrainName <span class="ot">&lt;-</span> <span class="st">&quot;celltype.l2&quot;</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>  PhiRes <span class="ot">&lt;-</span> <span class="fu">PhiSpaceR_1ref</span>(</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>    reference, </span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a>    <span class="at">query =</span> bridgeRNA, </span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a>    <span class="at">phenotypes =</span> YtrainName, </span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a>    <span class="at">PhiSpaceAssay =</span> PhiSpaceAssay,</span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a>    <span class="at">regMethod =</span> <span class="st">&quot;PLS&quot;</span>,</span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a>    <span class="at">scale =</span> <span class="cn">FALSE</span></span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a>  )</span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a>  <span class="fu">saveRDS</span>(PhiRes, PhiResPath)</span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a>  PhiRes <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(PhiResPath)</span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a>}</span></code></pre></div>
<p>Note that the bridge dataset has now been annotated.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>bridgeAnn <span class="ot">&lt;-</span> <span class="fu">normPhiScores</span>(PhiRes<span class="sc">$</span>PhiSpaceScore)</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="fu">head</span>(bridgeAnn)</span></code></pre></div>
<pre><code>##                               Prog_RBC        gdT    CD4 Naive  CD4 Memory
## TTCGCAACAATAATGG-14-s4d8 -0.1707345986  0.3142937  0.024318290 -0.13258113
## TATCCAGCAATTGAGA-14-s4d8 -0.0539900171  0.1223587 -0.079132130  0.25909237
## ACTCGCGCAAACTGTT-14-s4d8  0.0136845293 -0.1913396  0.074089946  0.01314921
## GACTATTCATGTCGCG-14-s4d8  0.1381644903 -0.1472680 -0.139126446 -0.11354641
## CTAATGTCATTGTTGG-14-s4d8  0.0006747695 -0.1353683 -0.003968247  0.34867338
## TGACCAAGTAGACAAA-14-s4d8 -0.1286231148  0.1834022 -0.300778458 -0.01186283
##                            CD14 Mono      Naive B    CD8 Naive        Treg
## TTCGCAACAATAATGG-14-s4d8 -0.03508110  0.177299624 -0.005793706 -0.18729220
## TATCCAGCAATTGAGA-14-s4d8  0.03920407  0.070065045  0.069233769  0.02057322
## ACTCGCGCAAACTGTT-14-s4d8  0.55715881 -0.004419651 -0.146966721  0.12118794
## GACTATTCATGTCGCG-14-s4d8  0.01349943  0.657490769  0.010944475 -0.01029004
## CTAATGTCATTGTTGG-14-s4d8 -0.02409488  0.015744293 -0.075164530  0.22040510
## TGACCAAGTAGACAAA-14-s4d8  0.21756821  0.081154893  0.405585859  0.05267516
##                          CD8 Effector_2          NK         GMP CD8 Effector_1
## TTCGCAACAATAATGG-14-s4d8    -0.19359786 -0.23114732  0.10532532    0.030164486
## TATCCAGCAATTGAGA-14-s4d8     0.10336668 -0.04702014 -0.01607817   -0.009467531
## ACTCGCGCAAACTGTT-14-s4d8     0.02719140 -0.06959604 -0.08155136    0.009924969
## GACTATTCATGTCGCG-14-s4d8     0.18685296 -0.10031176  0.03269648   -0.083924599
## CTAATGTCATTGTTGG-14-s4d8    -0.06730475  0.04239765 -0.02165142    0.161820780
## TGACCAAGTAGACAAA-14-s4d8    -0.25069804  0.32892606 -0.22216454    0.247644301
##                              CD16 Mono         pDC CD8 Memory_1        MAIT
## TTCGCAACAATAATGG-14-s4d8  0.0267491818  0.07519187   0.10838891  0.15693329
## TATCCAGCAATTGAGA-14-s4d8  0.0004520518  0.04187673   0.16961830  0.28565867
## ACTCGCGCAAACTGTT-14-s4d8  0.0401469480  0.07347996   0.03591260 -0.11384447
## GACTATTCATGTCGCG-14-s4d8  0.0956767357  0.08542073  -0.11933952  0.15696125
## CTAATGTCATTGTTGG-14-s4d8  0.0896238445 -0.03629314   0.09045062 -0.32292449
## TGACCAAGTAGACAAA-14-s4d8 -0.2205784394  0.06602202   0.03483928 -0.09627815
##                              Memory B        cDC2 CD56 bright NK    Prog_B 2
## TTCGCAACAATAATGG-14-s4d8 -0.018326246 -0.14549994     0.10604594  0.67305697
## TATCCAGCAATTGAGA-14-s4d8 -0.157773873 -0.03388604     0.11796970  0.08840598
## ACTCGCGCAAACTGTT-14-s4d8 -0.011452452 -0.02233130     0.05531265 -0.01953300
## GACTATTCATGTCGCG-14-s4d8 -0.009060715 -0.15772452    -0.19432405 -0.03411201
## CTAATGTCATTGTTGG-14-s4d8 -0.087183069  0.02049838    -0.07045638 -0.03339777
## TGACCAAGTAGACAAA-14-s4d8 -0.034366419  0.04425579     0.04903824 -0.01442982
##                              Prog_Mk CD8 Memory_2   Plasmablast          HSC
## TTCGCAACAATAATGG-14-s4d8 -0.05306825   0.11360104 -0.0368015817 -0.012091183
## TATCCAGCAATTGAGA-14-s4d8 -0.23100435  -0.01249257 -0.0136314348 -0.128567181
## ACTCGCGCAAACTGTT-14-s4d8  0.20827609   0.04596609  0.0014879483  0.153315573
## GACTATTCATGTCGCG-14-s4d8  0.02229154  -0.04417593 -0.0519034938 -0.008114113
## CTAATGTCATTGTTGG-14-s4d8  0.24471407   0.12718647  0.0486055125  0.077611859
## TGACCAAGTAGACAAA-14-s4d8 -0.37960088   0.28419243 -0.0006143916 -0.182652596
##                                  LMPP     Prog_DC    Prog_B 1
## TTCGCAACAATAATGG-14-s4d8  0.382422573  0.54725478  0.52730214
## TATCCAGCAATTGAGA-14-s4d8 -0.072865706  0.06203403  0.04468466
## ACTCGCGCAAACTGTT-14-s4d8  0.036138028 -0.06053255  0.12502706
## GACTATTCATGTCGCG-14-s4d8 -0.113021357 -0.16087069  0.01583117
## CTAATGTCATTGTTGG-14-s4d8 -0.005820409 -0.07928313  0.06925808
## TGACCAAGTAGACAAA-14-s4d8  0.001141363  0.03850285 -0.13583515</code></pre>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">rm</span>(PhiRes); <span class="fu">gc</span>()</span></code></pre></div>
<pre><code>##             used   (Mb) gc trigger   (Mb) limit (Mb)  max used   (Mb)
## Ncells   6948578  371.1   12295371  656.7         NA   9705296  518.4
## Vcells 451979864 3448.4  581547160 4436.9      18432 465074601 3548.3</code></pre>
<p>Next we transfer this continuous annotation from bridge to query.
Note that we use bridgeATACpeaks as reference in the following code to
match the query. Peaks in both bridge and query have been normalised
using <a href="https://stuartlab.org/signac/articles/pbmc_vignette">TF-IDF</a>.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a>PhiResPath <span class="ot">&lt;-</span> <span class="fu">paste0</span>(dat_dir, <span class="st">&quot;output/PhiResATACpeaks.rds&quot;</span>)</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">file.exists</span>(PhiResPath)){</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>  PhiSpaceAssay <span class="ot">&lt;-</span> <span class="st">&quot;data&quot;</span></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>  PhiRes <span class="ot">&lt;-</span> <span class="fu">PhiSpaceR_1ref</span>(</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a>    <span class="at">reference =</span> bridgeATACpeaks, </span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a>    <span class="at">query =</span> queryPeaks, </span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a>    <span class="at">response =</span> bridgeAnn,</span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a>    <span class="at">PhiSpaceAssay =</span> PhiSpaceAssay,</span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a>    <span class="at">regMethod =</span> <span class="st">&quot;PLS&quot;</span>, </span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a>    <span class="at">center =</span> F,</span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a>    <span class="at">scale =</span> <span class="cn">FALSE</span></span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a>  )</span>
<span id="cb8-13"><a href="#cb8-13" tabindex="-1"></a>  <span class="fu">saveRDS</span>(PhiRes, PhiResPath)</span>
<span id="cb8-14"><a href="#cb8-14" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb8-15"><a href="#cb8-15" tabindex="-1"></a>  PhiRes <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(PhiResPath)</span>
<span id="cb8-16"><a href="#cb8-16" tabindex="-1"></a>}</span>
<span id="cb8-17"><a href="#cb8-17" tabindex="-1"></a>PhiScPeaks <span class="ot">&lt;-</span> <span class="fu">normPhiScores</span>(PhiRes<span class="sc">$</span>PhiSpaceScore)</span></code></pre></div>
<p>Note that we set <code>center = F</code> here since centring would
turn a sparse peak matrix to dense, consuming a lot of RAM. If you have
enough RAM, centring is still encouraged.</p>
<p>Alternatively, we can transfer continuous annotation via aggregated
peaks or gene activity (GA) score.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>PhiResPath <span class="ot">&lt;-</span> <span class="fu">paste0</span>(dat_dir, <span class="st">&quot;output/PhiResATAC_GA.rds&quot;</span>)</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="cf">if</span>(<span class="sc">!</span><span class="fu">file.exists</span>(PhiResPath)){</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>  PhiSpaceAssay <span class="ot">&lt;-</span> <span class="st">&quot;logcounts&quot;</span></span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a>  PhiRes <span class="ot">&lt;-</span> <span class="fu">PhiSpaceR_1ref</span>(</span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a>    <span class="at">reference =</span> bridgeRNA, </span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a>    <span class="at">query =</span> queryGA, </span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a>    <span class="at">response =</span> bridgeAnn,</span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a>    <span class="at">PhiSpaceAssay =</span> PhiSpaceAssay,</span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a>    <span class="at">regMethod =</span> <span class="st">&quot;PLS&quot;</span>, </span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a>    <span class="at">center =</span> T,</span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a>    <span class="at">scale =</span> <span class="cn">FALSE</span></span>
<span id="cb9-12"><a href="#cb9-12" tabindex="-1"></a>  )</span>
<span id="cb9-13"><a href="#cb9-13" tabindex="-1"></a>  <span class="fu">saveRDS</span>(PhiRes, PhiResPath)</span>
<span id="cb9-14"><a href="#cb9-14" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb9-15"><a href="#cb9-15" tabindex="-1"></a>  PhiRes <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(PhiResPath)</span>
<span id="cb9-16"><a href="#cb9-16" tabindex="-1"></a>}</span>
<span id="cb9-17"><a href="#cb9-17" tabindex="-1"></a>PhiScGA <span class="ot">&lt;-</span> <span class="fu">normPhiScores</span>(PhiRes<span class="sc">$</span>PhiSpaceScore)</span></code></pre></div>
<p>Now we can compare which annotation gave more accurate cell type
annotation. An immediate problem is that the ground truth cell types for
query were different from the reference cell types. Finding one-to-one
correspondence between them is tricky. However, it’s possible to define
broad cell types and align both sets of cell types with the broad cell
types. This will also allow us to compare the accuracy of the
peaks-based and GA-based approaches at the broad cell type level.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a>refTypes_l2 <span class="ot">&lt;-</span> <span class="fu">unique</span>(reference<span class="sc">$</span>celltype.l2) <span class="sc">%&gt;%</span> sort</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>refTypes_l1 <span class="ot">&lt;-</span> <span class="fu">sapply</span>(</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>  refTypes_l2,</span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a>  <span class="cf">function</span>(x){</span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a>    <span class="fu">unique</span>(reference<span class="sc">$</span>celltype.l1[reference<span class="sc">$</span>celltype.l2 <span class="sc">==</span> x])</span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a>  }</span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a>)</span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a>ref_lookup <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">l2 =</span> refTypes_l2, <span class="at">l1 =</span> refTypes_l1)</span>
<span id="cb10-9"><a href="#cb10-9" tabindex="-1"></a><span class="co"># Query original annotations (ILC is hard to classify)</span></span>
<span id="cb10-10"><a href="#cb10-10" tabindex="-1"></a>cellTypeTable <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(dat_dir, <span class="st">&quot;data/CellTypeTable.rds&quot;</span>))</span>
<span id="cb10-11"><a href="#cb10-11" tabindex="-1"></a>queryTypes_l2 <span class="ot">&lt;-</span> <span class="fu">rownames</span>(cellTypeTable) <span class="sc">%&gt;%</span> sort</span>
<span id="cb10-12"><a href="#cb10-12" tabindex="-1"></a>queryTypes_l1 <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb10-13"><a href="#cb10-13" tabindex="-1"></a>  <span class="st">&quot;B cell&quot;</span>, <span class="st">&quot;Mono/DC&quot;</span>, <span class="st">&quot;Mono/DC&quot;</span>, <span class="st">&quot;T cell&quot;</span>, <span class="st">&quot;T cell&quot;</span>,</span>
<span id="cb10-14"><a href="#cb10-14" tabindex="-1"></a>  <span class="st">&quot;T cell&quot;</span>, <span class="st">&quot;T cell&quot;</span>, <span class="st">&quot;Mono/DC&quot;</span>, <span class="st">&quot;Progenitor cells&quot;</span>, <span class="st">&quot;Progenitor cells&quot;</span>,</span>
<span id="cb10-15"><a href="#cb10-15" tabindex="-1"></a>  <span class="st">&quot;Progenitor cells&quot;</span>, <span class="st">&quot;Progenitor cells&quot;</span>, <span class="st">&quot;ILC&quot;</span>, <span class="st">&quot;Progenitor cells&quot;</span>, <span class="st">&quot;Progenitor cells&quot;</span>,</span>
<span id="cb10-16"><a href="#cb10-16" tabindex="-1"></a>  <span class="st">&quot;B cell&quot;</span>, <span class="st">&quot;NK&quot;</span>, <span class="st">&quot;Progenitor cells&quot;</span>, <span class="st">&quot;Mono/DC&quot;</span>, <span class="st">&quot;B cell&quot;</span>,</span>
<span id="cb10-17"><a href="#cb10-17" tabindex="-1"></a>  <span class="st">&quot;Progenitor cells&quot;</span>, <span class="st">&quot;B cell&quot;</span></span>
<span id="cb10-18"><a href="#cb10-18" tabindex="-1"></a>)</span>
<span id="cb10-19"><a href="#cb10-19" tabindex="-1"></a>query_lookup <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb10-20"><a href="#cb10-20" tabindex="-1"></a>  <span class="at">l2 =</span> queryTypes_l2,</span>
<span id="cb10-21"><a href="#cb10-21" tabindex="-1"></a>  <span class="at">l1 =</span> queryTypes_l1</span>
<span id="cb10-22"><a href="#cb10-22" tabindex="-1"></a>)</span></code></pre></div>
<p>Evaluate performances: calculating overall and balanced (per cell
type average) classification errors. Indeed using aggregated peaks gave
lower erros compared to using peaks.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a>originAnn <span class="ot">&lt;-</span> queryGA<span class="sc">$</span>cellType</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>originAnn_l1 <span class="ot">&lt;-</span> query_lookup<span class="sc">$</span>l1[<span class="fu">match</span>(originAnn, query_lookup<span class="sc">$</span>l2)]</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="co"># Translate PhiSpace labels</span></span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>PhiSpaceAnn <span class="ot">&lt;-</span> PhiScGA <span class="sc">%&gt;%</span> getClass</span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a>PhiSpaceAnn_l1 <span class="ot">&lt;-</span> ref_lookup<span class="sc">$</span>l1[<span class="fu">match</span>(PhiSpaceAnn, ref_lookup<span class="sc">$</span>l2)]</span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a>PhiSpaceAnnPeaks <span class="ot">&lt;-</span> PhiScPeaks <span class="sc">%&gt;%</span> <span class="fu">getClass</span>()</span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a>PhiSpaceANN_l1_peaks <span class="ot">&lt;-</span> ref_lookup<span class="sc">$</span>l1[<span class="fu">match</span>(PhiSpaceAnnPeaks, ref_lookup<span class="sc">$</span>l2)]</span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a>PhiSpace<span class="sc">:::</span><span class="fu">classErr</span>(PhiSpaceAnn_l1, originAnn_l1)<span class="sc">$</span>err</span></code></pre></div>
<pre><code>## [1] 0.2291131 0.3536215</code></pre>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a>PhiSpace<span class="sc">:::</span><span class="fu">classErr</span>(PhiSpaceANN_l1_peaks, originAnn_l1)<span class="sc">$</span>err</span></code></pre></div>
<pre><code>## [1] 0.2892031 0.3978839</code></pre>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
