---
title: "Annotation of scATAC-seq using scRNA-seq reference"
author: "Jiadong Mao"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Annotation of scATAC-seq using scRNA-seq reference}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
# knitr::opts_chunk$set(eval = FALSE)
```

## Highlights

- ScRNA-seq reference + scATAC-seq query

- Utilising a bimodal bridge dataset

- Comprehensive benchmark study


# Introduction
Single-cell ATAC-seq (Assay for Transposase-Accessible Chromatin sequencing) measures chromatin accessibility across the genome. In the nucleus of each cell, each DNA chain is like a long thread wrapping around protein spools called histones. Each DNA-wrapped histone is known as a nucleosome -- the basic unit of chromatin. Measuring the openness of chromatin regions tells us two things:

- Whether a gene can be transcribed. For a gene to be transcribed, the gene's corresponding chromatin region has to be open, i.e. part of the DNA chain being unwound. 

- Transcription factor (TF) footprints. TFs bind to gene regions to promote or suppress gene expression. ATAC-seq can infer which regions TFs are binding to.

ScATAC-seq allows us to infer cell-type-specific gene regulations. Hence identifying cell types based on scATAC-seq is an important step. One common appraoch is to view ATAC-seq as surrogate for RNA-seq. That is, if a gene has more open regions, then we assume that it has more transcripts.

In this case study, we use a reference-based approach to annotate cell types in scATAC-seq. This approach is inspired by bridge integration implemented in Seurat V5 ([Hao et al., 2024](https://www.nature.com/articles/s41587-023-01767-y)). The idea is to transfer cell type annotations in scRNA-seq reference to scATAC-seq query, via a bimodal bridge dataset with matchcd RNA and ATAC features (e.g. generated from 10x multiome).

We use 

- Reference: BMMC (bone marrow mononuclear cell) scRNA-seq data set ([Stuart et al. (2019)](https://www.cell.com/cell/fulltext/S0092-8674(19)30559-8?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS0092867419305598%3Fshowall%3Dtrue));

- Bridge and query: 10x multiome (scRNA+ATAC-seq) dataset with 13 batches from [Leucken et al. (2021)](https://datasets-benchmarks-proceedings.neurips.cc/paper/2021/hash/158f3069a435b314a80bdcb024f8e422-Abstract-round2.html).

In our [paper](https://www.biorxiv.org/content/10.1101/2024.06.19.599787v1.full) we used a cross-validation (CV) scheme as follows. Each time we use one of the 13 batches of multiome data as the bridge dataset to facilitate the transfer of cell type annotations from the scRNA-seq reference to the remaining 12 batches of query scATAC-seq data (with their RNA part and ground truth cell type labels hidden). This allowed us to benchmark PhiSpace cross-modality annotation with Seurat V5. Here we simply show how PhiSpace bridge annotation can be done using one bridge and one query datasets.


All data related to this case study can be downloaded [here](https://www.dropbox.com/scl/fo/jeuqzjfyyr2j7doa922ve/AI-2U_wtZBpPGOswxMzReGQ?rlkey=n328yyr2llf81gz3chynjg0r6&st=nb1svufg&dl=0).
```{r}
suppressPackageStartupMessages(library(PhiSpace))
suppressPackageStartupMessages(library(vizOmics))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(magrittr))
suppressPackageStartupMessages(library(ggpubr))
suppressPackageStartupMessages(library(tidyr))
suppressPackageStartupMessages(library(RColorBrewer))

dat_dir <- "~/Dropbox/Research_projects/PhiSpace/VignetteData/ATAC/" # Replace this by your own directory 
source("~/Dropbox/Research_projects/PhiSpace/VignetteData/ATAC/utils.R")
```

# Peaks or aggregated peaks?

ScATAC-seq data are often available in the peak by cell format, where a peak refers to a small region on the genome that are open in a significant number of cells (i.e. region of interest). Peaking calling, i.e. defining the peaks based on the raw sequencing data, is a sophisticated computational step and there are a few different approaches, which we cannot cover here. Peaks are often further aggregated to the gene level as a proxy of gene expression. That is, open peaks falling in the region of a certain gene are summed up as 'gene activity score', approximating the expression level of that gene. 

In our paper, we demonstrated that gene activite scores (aggregted peaks) are more suitable for cell typing due to its enhanced data quality. Here we show PhiSpace cross-modality cell typing using both peaks and gene activity scores.

```{r}
reference <- readRDS(paste0(dat_dir, "data/obj.rna_for_refMap_sce.rds"))
bridgeRNA <- readRDS(paste0(dat_dir, "data/bridgeRNA_s4d8.rds"))
bridgeATACpeaks <- readRDS(paste0(dat_dir, "data/bridgeATACpeaks_s4d8.rds"))
bridgeGA <- readRDS(paste0(dat_dir, "data/bridgeGA_s4d8.rds"))
queryPeaks <- readRDS(paste0(dat_dir, "data/queryATACpeaks_s1d1.rds"))
queryGA <- readRDS(paste0(dat_dir, "data/queryATAC_GA_s1d1.rds"))
```


Run PhiSpace to transfer annotations from reference scRNA-seq to bridge scRNA-seq (RNA part of the bimodal bridge dataset).
```{r}
PhiResPath <- paste0(dat_dir, "output/PhiResRNA.rds")
if(!file.exists(PhiResPath)){
  PhiSpaceAssay <- "logcounts"
  YtrainName <- "celltype.l2"
  PhiRes <- PhiSpaceR_1ref(
    reference, 
    query = bridgeRNA, 
    phenotypes = YtrainName, 
    refAssay = PhiSpaceAssay,
    regMethod = "PLS",
    scale = FALSE
  )
  saveRDS(PhiRes, PhiResPath)
} else {
  PhiRes <- readRDS(PhiResPath)
}
```

Note that the bridge dataset has now been annotated.
```{r}
bridgeAnn <- normPhiScores(PhiRes$PhiSpaceScore)
head(bridgeAnn)
rm(PhiRes); gc()
```


Next we transfer this continuous annotation from bridge to query. Note that we use bridgeATACpeaks as reference in the following code to match the query. Peaks in both bridge and query have been normalised using [TF-IDF](https://stuartlab.org/signac/articles/pbmc_vignette).
```{r}
PhiResPath <- paste0(dat_dir, "output/PhiResATACpeaks.rds")
if(!file.exists(PhiResPath)){
  PhiSpaceAssay <- "data"
  PhiRes <- PhiSpaceR_1ref(
    reference = bridgeATACpeaks, 
    query = queryPeaks, 
    response = bridgeAnn,
    refAssay = PhiSpaceAssay,
    regMethod = "PLS", 
    center = F,
    scale = FALSE
  )
  saveRDS(PhiRes, PhiResPath)
} else {
  PhiRes <- readRDS(PhiResPath)
}
PhiScPeaks <- normPhiScores(PhiRes$PhiSpaceScore)
```

Note that we set `center = F` here since centring would turn a sparse peak matrix to dense, consuming a lot of RAM. If you have enough RAM, centring is still encouraged.


Alternatively, we can transfer continuous annotation via aggregated peaks or gene activity (GA) score.
```{r}
PhiResPath <- paste0(dat_dir, "output/PhiResATAC_GA.rds")
if(!file.exists(PhiResPath)){
  PhiSpaceAssay <- "logcounts"
  PhiRes <- PhiSpaceR_1ref(
    reference = bridgeRNA, 
    query = queryGA, 
    response = bridgeAnn,
    refAssay = PhiSpaceAssay,
    regMethod = "PLS", 
    center = T,
    scale = FALSE
  )
  saveRDS(PhiRes, PhiResPath)
} else {
  PhiRes <- readRDS(PhiResPath)
}
PhiScGA <- normPhiScores(PhiRes$PhiSpaceScore)
```



Now we can compare which annotation gave more accurate cell type annotation. An immediate problem is that the ground truth cell types for query were different from the reference cell types. Finding one-to-one correspondence between them is tricky. However, it's possible to define broad cell types and align both sets of cell types with the broad cell types. This will also allow us to compare the accuracy of the peaks-based and GA-based approaches at the broad cell type level.
```{r}
refTypes_l2 <- unique(reference$celltype.l2) %>% sort
refTypes_l1 <- sapply(
  refTypes_l2,
  function(x){
    unique(reference$celltype.l1[reference$celltype.l2 == x])
  }
)
ref_lookup <- data.frame(l2 = refTypes_l2, l1 = refTypes_l1)
# Query original annotations (ILC is hard to classify)
cellTypeTable <- readRDS(paste0(dat_dir, "data/CellTypeTable.rds"))
queryTypes_l2 <- rownames(cellTypeTable) %>% sort
queryTypes_l1 <- c(
  "B cell", "Mono/DC", "Mono/DC", "T cell", "T cell",
  "T cell", "T cell", "Mono/DC", "Progenitor cells", "Progenitor cells",
  "Progenitor cells", "Progenitor cells", "ILC", "Progenitor cells", "Progenitor cells",
  "B cell", "NK", "Progenitor cells", "Mono/DC", "B cell",
  "Progenitor cells", "B cell"
)
query_lookup <- data.frame(
  l2 = queryTypes_l2,
  l1 = queryTypes_l1
)
```

Evaluate performances: calculating overall and balanced (per cell type average) classification errors. Indeed using aggregated peaks gave lower erros compared to using peaks.
```{r}
originAnn <- queryGA$cellType
originAnn_l1 <- query_lookup$l1[match(originAnn, query_lookup$l2)]
# Translate PhiSpace labels
PhiSpaceAnn <- PhiScGA %>% getClass
PhiSpaceAnn_l1 <- ref_lookup$l1[match(PhiSpaceAnn, ref_lookup$l2)]
PhiSpaceAnnPeaks <- PhiScPeaks %>% getClass()
PhiSpaceANN_l1_peaks <- ref_lookup$l1[match(PhiSpaceAnnPeaks, ref_lookup$l2)]
PhiSpace:::classErr(PhiSpaceAnn_l1, originAnn_l1)$err
PhiSpace:::classErr(PhiSpaceANN_l1_peaks, originAnn_l1)$err
```











