<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Cell state annotation for single-cell and spatial omics data. • PhiSpace</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="48x48" href="favicon-48x48.png">
<link rel="icon" type="”image/svg+xml”" href="favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png">
<link rel="icon" sizes="any" href="favicon.ico">
<link rel="manifest" href="site.webmanifest">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Cell state annotation for single-cell and spatial omics data.">
<meta property="og:description" content="Continous phenotyping of bulk, single-cell and spatial omics data using partial least squares regression.">
<meta property="og:image" content="https://jiadongm.github.io/PhiSpace/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">PhiSpace</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>

  </a>
</li>
<li>
  <a href="articles/index.html">Vignettes</a>
</li>
<li>
  <a href="reference/index.html">Functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="φ-space-continuous-cell-state-annotation-for-single-cell-and-spatial-omics-studies">Φ-Space: continuous cell state annotation for single-cell and spatial omics studies<a class="anchor" aria-label="anchor" href="#%CF%86-space-continuous-cell-state-annotation-for-single-cell-and-spatial-omics-studies"></a>
</h1></div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>Installing Bioconductor dependencies and then PhiSpace:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">'jiadongm/PhiSpace/pkg'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="what-is-φ-space">What is Φ-Space?<a class="anchor" aria-label="anchor" href="#what-is-%CF%86-space"></a>
</h2>
<p>Φ-Space is a computational framework for reference-based continuous annotation of cell states in single-cell and spatial multiomics data. Currently it has two modules:</p>
<ul>
<li>PhiSpace multiomics for single-cell multiomics data (<a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-025-03755-8" class="external-link">Mao et al., 2024</a>), and</li>
<li>PhiSpace ST for spatial transcriptomics data (<a href="https://www.biorxiv.org/content/10.1101/2024.06.19.599787v1" class="external-link">Mao et al., 2025</a>)</li>
</ul>
<p>Given a bulk or single-cell RNA-seq <strong>reference</strong> dataset with potentially multiple layers of phenotypes defined in the metadata (e.g. cell type and sample source), Φ-Space can phenotype on a continuum the cells and cell-like objects in the <strong>query</strong> dataset. The core of Φ-Space is continuous phenotyping based on partial least squares (PLS) regression. Compared to conventional cell type annotation methods, Φ-Space has the following strengthss</p>
<ul>
<li>Identifying continuous and out-of-reference cell states;</li>
<li>Robust against batch effects;</li>
<li>Utilising bulk and multiomic refereneces and queries;</li>
<li>More suitable for exploring the spatial patterns of rare cell types.</li>
</ul>
<p>We have applied Φ-Space to many different use cases, including</p>
<table class="table">
<colgroup>
<col width="25%">
<col width="25%">
<col width="25%">
<col width="25%">
</colgroup>
<thead><tr>
<th align="left">Reference</th>
<th align="left">Query</th>
<th align="left">Note</th>
<th align="left">Vignettes</th>
</tr></thead>
<tbody>
<tr>
<td align="left">bulk RNA-seq</td>
<td align="left">scRNA-seq</td>
<td align="left">Dendritic cell development</td>
<td align="left"><a href="articles/getting_started.html">Stemformatics DC atlas</a></td>
</tr>
<tr>
<td align="left">bulk RNA-seq</td>
<td align="left">scRNA-seq (Perturb-seq)</td>
<td align="left">T cell states development</td>
<td align="left"><a href="articles/PerturbSeq.html">Perturb-seq</a></td>
</tr>
<tr>
<td align="left">scRNA-seq</td>
<td align="left">scRNA-seq</td>
<td align="left"></td>
<td align="left">TBC</td>
</tr>
<tr>
<td align="left">scRNA-seq</td>
<td align="left">scATAC-seq</td>
<td align="left">requires a bimodal bridge dataset</td>
<td align="left"><a href="articles/BridgeAnnotation.html">Bridge annotation</a></td>
</tr>
<tr>
<td align="left">CITE-seq (scRNA+Protein-seq)</td>
<td align="left">CITE-seq</td>
<td align="left">using both modalities</td>
<td align="left"><a href="articles/CITE-seq.html">Covid CITE-seq</a></td>
</tr>
<tr>
<td align="left">scRNA-seq</td>
<td align="left">supercellular spatial RNA-seq</td>
<td align="left">e.g. 10x Visium, Slide-seqV2</td>
<td align="left"><a href="articles/Visium.html">Cell type deconvolution for Visium</a></td>
</tr>
<tr>
<td align="left">scRNA-seq</td>
<td align="left">subcellular spatial RNA-seq (imaging-based)</td>
<td align="left">e.g. CosMx, 10x Xenium</td>
<td align="left"><a href="articles/CosMx.html">CosMx lung cancer microenvironment</a></td>
</tr>
<tr>
<td align="left">scRNA-seq</td>
<td align="left">subcellular spatial RNA-seq (sequencing-based)</td>
<td align="left">e.g. Stereo-seq, 10x Visium HD</td>
<td align="left"><a href="articles/StereoSeq.html">Stereo-seq AML mouse spleen with cancer lieage tracing</a></td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="cite-phispace">Cite PhiSpace<a class="anchor" aria-label="anchor" href="#cite-phispace"></a>
</h2>
<p>Mao, Jiadong, Choi, Jarny and Lê Cao, Kim-Anh. (2025). Φ-Space ST: a platform-agnostic method to identify cell states in spatial transcriptomics studies. <a href="https://www.biorxiv.org/content/10.1101/2025.02.05.636735v1" class="external-link"><em>bioRxiv</em></a>.</p>
<p>Mao, Jiadong, Deng, Yidi and Lê Cao, Kim-Anh. (2024). Φ-Space: Continuous phenotyping of single-cell multi-omics data. <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-025-03755-8" class="external-link"><em>Genome Biology</em></a>.</p>
<p>Check out our talk for PhiSpace single-cell multiomics at <a href="http://mixomics.org/2024/06/phispace/" class="external-link">mixOmics website</a>.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing PhiSpace</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="https://github.com/jiadongm" class="external-link">Jiadong Mao</a> <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-3818-1981" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://github.com/jiadongm" class="external-link">Jiadong Mao</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
