<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ PhiSpace</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"><meta property="og:image" content="https://jiadongm.github.io/PhiSpace/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">PhiSpace</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/getting_started.html">Getting Started: mapping scRNA-seq to bulk RNA-seq atlas</a></li>
    <li><a class="dropdown-item" href="../articles/CITE-seq.html">Multiomic annotation: using CITE-seq reference to annotate CITE-seq query</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="CLRnorm.html">CLRnorm()</a></code>

        </dt>
        <dd>Center log ratio normalisation</dd>
      </dl><dl><dt>

          <code><a href="CVTune_ncomp.html">CVTune_ncomp()</a></code>

        </dt>
        <dd>Tune ncomp using cross-validation.</dd>
      </dl><dl><dt>

          <code><a href="CVTune_nfeat.html">CVTune_nfeat()</a></code>

        </dt>
        <dd>Tune nfeat using cross-validation.</dd>
      </dl><dl><dt>

          <code><a href="KeepCommonGenes.html">KeepCommonGenes()</a></code>

        </dt>
        <dd>Keep common genes of two SingleCellExperiment objects.</dd>
      </dl><dl><dt>

          <code><a href="PhiSpace.html">PhiSpace()</a></code>

        </dt>
        <dd>PhiSpace continuous phenotyping</dd>
      </dl><dl><dt>

          <code><a href="PhiSpaceR_1ref.html">PhiSpaceR_1ref()</a></code>

        </dt>
        <dd>PhiSpace using a single reference</dd>
      </dl><dl><dt>

          <code><a href="RTassay.html">RTassay()</a></code>

        </dt>
        <dd>Apply rank transform to a gene by cell matrix.</dd>
      </dl><dl><dt>

          <code><a href="RankTransf.html">RankTransf()</a></code>

        </dt>
        <dd>Rank transform of a <code>SingleCellExperiment</code> object.</dd>
      </dl><dl><dt>

          <code><a href="SuperPC.html">SuperPC()</a></code>

        </dt>
        <dd>Compute regression model using selected features.</dd>
      </dl><dl><dt>

          <code><a href="VizSpatial.html">VizSpatial()</a></code>

        </dt>
        <dd>Visualising features (eg gene expression), metadata (eg cell type) or reducedDim component (eg PhiSpace cell type score) in 2D tissue space.</dd>
      </dl><dl><dt>

          <code><a href="align_clusters.html">align_clusters()</a></code>

        </dt>
        <dd>Align clustering results.</dd>
      </dl><dl><dt>

          <code><a href="as.sparse.matrix.html">as.sparse.matrix()</a></code>

        </dt>
        <dd>Convert a matrix to sparse metrix.</dd>
      </dl><dl><dt>

          <code><a href="basicPlotFormat.html">basicPlotFormat()</a></code>

        </dt>
        <dd>Plot formatting for presentation and publication</dd>
      </dl><dl><dt>

          <code><a href="cellTypeCorMat.html">cellTypeCorMat()</a></code>

        </dt>
        <dd>Create cell type correlation matrices</dd>
      </dl><dl><dt>

          <code><a href="classErr.html">classErr()</a></code>

        </dt>
        <dd>Calculate classification errors.</dd>
      </dl><dl><dt>

          <code><a href="codeY.html">codeY()</a></code>

        </dt>
        <dd>Convert discrete annotations to dummy variable matrices.</dd>
      </dl><dl><dt>

          <code><a href="codeY_vec.html">codeY_vec()</a></code>

        </dt>
        <dd>Convert discrete annotations to dummy variable matrices.</dd>
      </dl><dl><dt>

          <code><a href="computUMAP.html">computUMAP()</a></code>

        </dt>
        <dd>Compute UMAP.</dd>
      </dl><dl><dt>

          <code><a href="doubleCent.html">doubleCent()</a></code>

        </dt>
        <dd>Double center a matrix by column and row means, resulting in a new matrix with zero row and column means.</dd>
      </dl><dl><dt>

          <code><a href="getClass.html">getClass()</a></code>

        </dt>
        <dd>Dicretise soft annotation to labels.</dd>
      </dl><dl><dt>

          <code><a href="getErr.html">getErr()</a></code>

        </dt>
        <dd>Compute different errors for cross-validation.</dd>
      </dl><dl><dt>

          <code><a href="getErr_nfeat.html">getErr_nfeat()</a></code>

        </dt>
        <dd>Calculate errors for the cross-validation for selecting nfeat.</dd>
      </dl><dl><dt>

          <code><a href="getPC.html">getPC()</a></code>

        </dt>
        <dd>Principal component analysis (PCA) based on partial singular value decomposition (SVD).</dd>
      </dl><dl><dt>

          <code><a href="loadBarplot.html">loadBarplot()</a></code>

        </dt>
        <dd>Plot loadings of dimension reduction as bar plot.</dd>
      </dl><dl><dt>

          <code><a href="logTransf.html">logTransf()</a></code>

        </dt>
        <dd>Relative counts normalisation.</dd>
      </dl><dl><dt>

          <code><a href="matrixPlot.html">matrixPlot()</a></code>

        </dt>
        <dd>Score matrix plot</dd>
      </dl><dl><dt>

          <code><a href="mvr.html">mvr()</a></code>

        </dt>
        <dd>Multivariate regression via principal component analysis (PCA) or partial least squares (PLS).</dd>
      </dl><dl><dt>

          <code><a href="normPhiScores.html">normPhiScores()</a></code>

        </dt>
        <dd>Normalise PhiSpace phenotypic scores by row or by column.</dd>
      </dl><dl><dt>

          <code><a href="phenotype.html">phenotype()</a></code>

        </dt>
        <dd>Soft phenotyping of query assay.</dd>
      </dl><dl><dt>

          <code><a href="plotPhiSpaceHeatMap.html">plotPhiSpaceHeatMap()</a></code>

        </dt>
        <dd>Plot PhiSpace scores as a heatmap</dd>
      </dl><dl><dt>

          <code><a href="plotSankey.html">plotSankey()</a></code>

        </dt>
        <dd>Plot Sankey diagram for two classification results.</dd>
      </dl><dl><dt>

          <code><a href="plotSankey3.html">plotSankey3()</a></code>

        </dt>
        <dd>Plot Sankey diagram for three hard classification results.</dd>
      </dl><dl><dt>

          <code><a href="pseudoBulk.html">pseudoBulk()</a></code>

        </dt>
        <dd>Converting single cells to pseudo-bulks.</dd>
      </dl><dl><dt>

          <code><a href="reNameCols.html">reNameCols()</a></code>

        </dt>
        <dd>Turn a matrix to a data.frame withe newly defined colnames.</dd>
      </dl><dl><dt>

          <code><a href="scranTransf.html">scranTransf()</a></code>

        </dt>
        <dd>Run scran normalisation</dd>
      </dl><dl><dt>

          <code><a href="selectFeat.html">selectFeat()</a></code>

        </dt>
        <dd>Select nfeat top features from each column (label) of impScores</dd>
      </dl><dl><dt>

          <code><a href="split2.html">split2()</a></code>

        </dt>
        <dd>Create partitions of data for cross-validation.</dd>
      </dl><dl><dt>

          <code><a href="subsample.html">subsample()</a></code>

        </dt>
        <dd>Subsample from a sce object</dd>
      </dl><dl><dt>

          <code><a href="translateLabel.html">translateLabel()</a></code>

        </dt>
        <dd>Translate cell type labels from finer to coarser according to a dictionary.</dd>
      </dl><dl><dt>

          <code><a href="tunePhiSpace.html">tunePhiSpace()</a></code>

        </dt>
        <dd>Title</dd>
      </dl><dl><dt>

          <code><a href="zeroFeatQC.html">zeroFeatQC()</a></code>

        </dt>
        <dd>Minimal quality control (QC): remove features with all zero values.</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jiadong Mao.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

