<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Package index â€¢ PhiSpace</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">PhiSpace</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="CLRnorm.html">CLRnorm()</a></code> </p>
        </td>
        <td><p>Center log ratio normalisation</p></td>
      </tr><tr><td>
          <p><code><a href="CVTune_ncomp.html">CVTune_ncomp()</a></code> </p>
        </td>
        <td><p>Tune ncomp using cross-validation.</p></td>
      </tr><tr><td>
          <p><code><a href="CVTune_nfeat.html">CVTune_nfeat()</a></code> </p>
        </td>
        <td><p>Tune nfeat using cross-validation.</p></td>
      </tr><tr><td>
          <p><code><a href="KeepCommonGenes.html">KeepCommonGenes()</a></code> </p>
        </td>
        <td><p>Keep common genes of two SingleCellExperiment objects.</p></td>
      </tr><tr><td>
          <p><code><a href="PhiSpace.html">PhiSpace()</a></code> </p>
        </td>
        <td><p>PhiSpace continuous phenotyping</p></td>
      </tr><tr><td>
          <p><code><a href="PhiSpaceR_1ref.html">PhiSpaceR_1ref()</a></code> </p>
        </td>
        <td><p>PhiSpace using a single reference</p></td>
      </tr><tr><td>
          <p><code><a href="RTassay.html">RTassay()</a></code> </p>
        </td>
        <td><p>Apply rank transform to a gene by cell matrix.</p></td>
      </tr><tr><td>
          <p><code><a href="RankTransf.html">RankTransf()</a></code> </p>
        </td>
        <td><p>Rank transform of a <code>SingleCellExperiment</code> object.</p></td>
      </tr><tr><td>
          <p><code><a href="SuperPC.html">SuperPC()</a></code> </p>
        </td>
        <td><p>Compute regression model using selected features.</p></td>
      </tr><tr><td>
          <p><code><a href="VizSpatial.html">VizSpatial()</a></code> </p>
        </td>
        <td><p>Visualising features (eg gene expression), metadata (eg cell type) or reducedDim component (eg PhiSpace cell type score) in 2D tissue space.</p></td>
      </tr><tr><td>
          <p><code><a href="align_clusters.html">align_clusters()</a></code> </p>
        </td>
        <td><p>Align clustering results.</p></td>
      </tr><tr><td>
          <p><code><a href="as.sparse.matrix.html">as.sparse.matrix()</a></code> </p>
        </td>
        <td><p>Convert a matrix to sparse metrix.</p></td>
      </tr><tr><td>
          <p><code><a href="basicPlotFormat.html">basicPlotFormat()</a></code> </p>
        </td>
        <td><p>Plot formatting for presentation and publication</p></td>
      </tr><tr><td>
          <p><code><a href="cellTypeCorMat.html">cellTypeCorMat()</a></code> </p>
        </td>
        <td><p>Create cell type correlation matrices</p></td>
      </tr><tr><td>
          <p><code><a href="classErr.html">classErr()</a></code> </p>
        </td>
        <td><p>Calculate classification errors.</p></td>
      </tr><tr><td>
          <p><code><a href="codeY.html">codeY()</a></code> </p>
        </td>
        <td><p>Convert discrete annotations to dummy variable matrices.</p></td>
      </tr><tr><td>
          <p><code><a href="codeY_vec.html">codeY_vec()</a></code> </p>
        </td>
        <td><p>Convert discrete annotations to dummy variable matrices.</p></td>
      </tr><tr><td>
          <p><code><a href="computUMAP.html">computUMAP()</a></code> </p>
        </td>
        <td><p>Compute UMAP.</p></td>
      </tr><tr><td>
          <p><code><a href="doubleCent.html">doubleCent()</a></code> </p>
        </td>
        <td><p>Double center a matrix by column and row means, resulting in a new matrix with zero row and column means.</p></td>
      </tr><tr><td>
          <p><code><a href="getClass.html">getClass()</a></code> </p>
        </td>
        <td><p>Dicretise soft annotation to labels.</p></td>
      </tr><tr><td>
          <p><code><a href="getErr.html">getErr()</a></code> </p>
        </td>
        <td><p>Compute different errors for cross-validation.</p></td>
      </tr><tr><td>
          <p><code><a href="getErr_nfeat.html">getErr_nfeat()</a></code> </p>
        </td>
        <td><p>Calculate errors for the cross-validation for selecting nfeat.</p></td>
      </tr><tr><td>
          <p><code><a href="getPC.html">getPC()</a></code> </p>
        </td>
        <td><p>Principal component analysis (PCA) based on partial singular value decomposition (SVD).</p></td>
      </tr><tr><td>
          <p><code><a href="loadBarplot.html">loadBarplot()</a></code> </p>
        </td>
        <td><p>Plot loadings of dimension reduction as bar plot.</p></td>
      </tr><tr><td>
          <p><code><a href="logTransf.html">logTransf()</a></code> </p>
        </td>
        <td><p>Relative counts normalisation.</p></td>
      </tr><tr><td>
          <p><code><a href="matrixPlot.html">matrixPlot()</a></code> </p>
        </td>
        <td><p>Score matrix plot</p></td>
      </tr><tr><td>
          <p><code><a href="mvr.html">mvr()</a></code> </p>
        </td>
        <td><p>Multivariate regression via principal component analysis (PCA) or partial least squares (PLS).</p></td>
      </tr><tr><td>
          <p><code><a href="normPhiScores.html">normPhiScores()</a></code> </p>
        </td>
        <td><p>Normalise PhiSpace phenotypic scores by row or by column.</p></td>
      </tr><tr><td>
          <p><code><a href="phenotype.html">phenotype()</a></code> </p>
        </td>
        <td><p>Soft phenotyping of query assay.</p></td>
      </tr><tr><td>
          <p><code><a href="plotPhiSpaceHeatMap.html">plotPhiSpaceHeatMap()</a></code> </p>
        </td>
        <td><p>Plot PhiSpace scores as a heatmap</p></td>
      </tr><tr><td>
          <p><code><a href="plotSankey.html">plotSankey()</a></code> </p>
        </td>
        <td><p>Plot Sankey diagram for two classification results.</p></td>
      </tr><tr><td>
          <p><code><a href="plotSankey3.html">plotSankey3()</a></code> </p>
        </td>
        <td><p>Plot Sankey diagram for three hard classification results.</p></td>
      </tr><tr><td>
          <p><code><a href="pseudoBulk.html">pseudoBulk()</a></code> </p>
        </td>
        <td><p>Converting single cells to pseudo-bulks.</p></td>
      </tr><tr><td>
          <p><code><a href="reNameCols.html">reNameCols()</a></code> </p>
        </td>
        <td><p>Turn a matrix to a data.frame withe newly defined colnames.</p></td>
      </tr><tr><td>
          <p><code><a href="scranTransf.html">scranTransf()</a></code> </p>
        </td>
        <td><p>Run scran normalisation</p></td>
      </tr><tr><td>
          <p><code><a href="selectFeat.html">selectFeat()</a></code> </p>
        </td>
        <td><p>Select nfeat top features from each column (label) of impScores</p></td>
      </tr><tr><td>
          <p><code><a href="split2.html">split2()</a></code> </p>
        </td>
        <td><p>Create partitions of data for cross-validation.</p></td>
      </tr><tr><td>
          <p><code><a href="subsample.html">subsample()</a></code> </p>
        </td>
        <td><p>Subsample from a sce object</p></td>
      </tr><tr><td>
          <p><code><a href="translateLabel.html">translateLabel()</a></code> </p>
        </td>
        <td><p>Translate cell type labels from finer to coarser according to a dictionary.</p></td>
      </tr><tr><td>
          <p><code><a href="tunePhiSpace.html">tunePhiSpace()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="zeroFeatQC.html">zeroFeatQC()</a></code> </p>
        </td>
        <td><p>Minimal quality control (QC): remove features with all zero values.</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Jiadong Mao.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

